rotkehlchen/exchanges/kucoin.py

```diff
@@ -0,0 +1,774 @@
+    400004: 'Invalid API passphrase.',
+    400005: 'Invalid API secret.',
+    400007: 'Provided KuCoin API key needs to have "General" permission activated.',
+    411100: 'Contact KuCoin support to unfreeze your account',
+}
+API_PAGE_SIZE_LIMIT = 500
+# Rate limit is 1800 requests per minute, exceed it multiple times the system
+# will restrict the IP
+API_REQUEST_RETRY_TIMES = 2
+API_REQUEST_RETRIES_AFTER_SECONDS = 1
+
+
+class KucoinCase(Enum):
+    API_KEY = 1
+    BALANCES = 2
+    TRADES = 3
+    DEPOSITS = 4
+    WITHDRAWALS = 5
+
+    def __str__(self) -> str:
+        if self == KucoinCase.API_KEY:
+            return 'api_key'
+        if self == KucoinCase.BALANCES:
+            return 'balances'
+        if self == KucoinCase.TRADES:
+            return 'trades'
+        if self == KucoinCase.DEPOSITS:
+            return 'deposits'
+        if self == KucoinCase.WITHDRAWALS:
+            return 'withdrawals'
+        raise AssertionError(f'Unexpected KucoinCase: {self}')
+
+
+class SkipReason(Enum):
+    AFTER_TIMESTAMP_RANGE = 1
+    BEFORE_TIMESTAMP_RANGE = 2
+    INNER_MOVEMENT = 3
+
+    def __str__(self) -> str:
+        if self == SkipReason.AFTER_TIMESTAMP_RANGE:
+            return 'after timestamp range'
+        if self == SkipReason.BEFORE_TIMESTAMP_RANGE:
+            return 'before timestamp range'
+        if self == SkipReason.INNER_MOVEMENT:
+            return 'that is an inner movement'
+        raise AssertionError(f'Unexpected SkipReason: {self}')
+
+
+DeserializationMethod = Callable[
+    ...,
+    Union[
+        Tuple[Optional[Trade], Optional[SkipReason]],
+        Tuple[Optional[AssetMovement], Optional[SkipReason]],
+    ],
+]
+
+
+def deserialize_trade_pair(trade_pair_symbol: str) -> TradePair:
+    """May raise:
+    - UnprocessableTradePair
+    - UnknownAsset
+    - UnsupportedAsset
+    """
+    try:
+        base_asset_symbol, quote_asset_symbol = trade_pair_symbol.split('-')
+    except ValueError as e:
+        raise UnprocessableTradePair(trade_pair_symbol) from e
+
+    base_asset = asset_from_kucoin(base_asset_symbol)
+    quote_asset = asset_from_kucoin(quote_asset_symbol)
+
+    return TradePair(f'{base_asset.identifier}_{quote_asset.identifier}')
+
+
+class Kucoin(ExchangeInterface):  # lgtm[py/missing-call-to-init]
+    """Resources:
+    https://docs.kucoin.com
+    https://github.com/Kucoin/kucoin-python-sdk
+    """
+    def __init__(
+            self,
+            api_key: ApiKey,
+            secret: ApiSecret,
+            database: 'DBHandler',
+            msg_aggregator: MessagesAggregator,
+            passphrase: str,
+            base_uri: str = 'https://api.kucoin.com',
+    ):
+        super().__init__(str(Location.KUCOIN), api_key, secret, database)
+        self.base_uri = base_uri
+        self.session.headers.update({
+            'Content-Type': 'application/json',
+            'KC-API-KEY': self.api_key,
+            'KC-API-PASSPHRASE': passphrase,
```

Comment:
@kelsos Please note this while integrating Kucoin. User needs to add a passphrase for this exchange.